### PR TITLE
[C++/ObjC] Add support for paths in front of struct/class identifiers

### PR DESCRIPTION
**Pull request by mjendruk**
I'm sorry for the new pull request. The old one was automatically closed for some reason.

## _Copied from old PR_:

When a struct or class has been forward declared in a namespace or other class, the class or struct can be defined using an identifier with a path in front of it. I added this possibility.

**Example**
```C++
namespace Namespace
{
  class BaseClass;
}

class Namespace::BaseClass {};
```

## _Last reply_:

Hi! Thank you all for your feedback. 

> @wbond The :: should not get entity.name.class, neither should the namespace. The namespace name should be something like support.namespace.

Yes that’s what I thought too and then I looked at how scopes got applied to function definitions and function calls with namespace paths and realised that `entity.name.function` and `variable.function` where applied to the entire path including the namespace. 

With my addition, indexing works as expected as the namespace gets stripped away in `Symbol Index.tmPreferences`.

### CODE REVIEW

File: C++/C++.sublime-syntax [hunk 1]
```diff
@@ -1044,6 +1050,12 @@ contexts:
     - match: '{{identifier}}(?={{data_structures_forward_decl_lookahead}})'
       scope: entity.name.struct.forward-decl.c++
       set: data-structures-struct-definition-after-identifier
+    - match: '(?={{path_lookahead}}\s*::\s*{{identifier}})'
+      set:
+        - meta_scope: entity.name.struct.c++
+        - include: identifiers
+        - match: '(?=[^\w\s])'
+          set: data-structures-struct-definition-after-identifier
```
Comment (rwols) [inline C++/C++.sublime-syntax]: I tested this out locally and it seems to work fine. Nice job! Only thing missing IMO are unions:

```
namespace ns {
class Bar;
struct Foo;
union Qux;
} // namespace ns

class ns::Bar {
public:
  Bar() = default;
};

struct ns::Foo {
  Foo() = default;
};

union ns::Qux {
  Qux() = default;
};
```

Comment (mjendruk) [reply C++/C++.sublime-syntax]: Okay the latest commit added support for both enums and unions. Should be good to go now.

**Comment by mjendruk** [comment]
Correct handling of macros is not yet working

```c++
class Namespace::MyClass Macro
{
};
```

**Comment by mjendruk** [comment]
There is still something weird though. In the code snippet below, both `Namespace` and `MyEnum` are recognized as symbols for `enum class Namespace::MyEnum` but for the other definition only `MyClass` is a symbol (as it should be).

```c++
class Namespace::MyClass
{
};

enum class Namespace::MyEnum
{
};
```

**Comment by wbond** [comment]
This will have to be adapted to that the `entity.name` is not applied to the namespace name, nor the `::`.

If somewhere else you are seeing such a pattern, I would consider that a bug that we need to fix also.

**Comment by mjendruk** [comment]
@wbond, yes I agree. My newer PRs #1819 and #1818 address this. This one needs to be updated.

**Comment by mjendruk** [comment]
Fixed it!

**Comment by wbond** [comment]
Thanks for these fixes, they look like solid enhancements!

**Pull request merged**